torch/_functorch/aot_autograd.py

```diff
@@ -422,6 +422,8 @@ class OutputAliasInfo:
     base_idx: Optional[int]
     # If it is a Tensor, what the dynamic dims are (otherwise is None)
     dynamic_dims: Optional[Set[int]]
+    # requires_grad
+    requires_grad: bool
```

Comment:
nit: `ViewAndMutationMeta.requires_grad_info` already has this info, since it is of length (# mutated inputs + # user outputs).

Reply:
Any reason you prefer to have a separate object rather than have it on here?

Reply:
Yeah this is a reasonable q: Let me give context. There are a few other places where we need to distinguish which tensors require gradients:

(1) Affects both mutated inputs and fw outputs ([code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L3121)): At runtime, the compiled forward graph returns **both** any updated inputs, and any user fw outs. For both of these groups of tensors, we need to know which ones do not require gradients, so we can mark them as non_differentiable.

(2) Affects just mutated inputs ([code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1177)): During tracing of the joint, if we have a mutated input that requires gradients, we need to clone() it in the forward, so that we can autograd.grad() w.r.t. the input pre-mutation (this is actually kind of sub-optimal since the clone() affects runtime perf and is just to appease the autograd engine, but this case should be rare). 

(3) Affects just fw outputs ([code](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L2685)): At runtime after our `CompiledFunction.apply()` finishes, we need to regenerate fw outs that alias inputs. We need to know if the output alias no longer requires grad, which implies that a detach() happened during tracing (so we need to reapply that detach at runtime).

If we do what you put above and move the requires_grad info directly on `OutputInfo`, then we'd have to either:

(a) Also put requires_grad info on `InputInfo` objects - **but**, every piece of code in AOTAutograd that care about requires grad information on inputs would have to manually filter down to the input infos that correspond to mutated inputs

(b) Add another piece of metadata that just correponds to "mutated inputs", that we tack requires_grad info on.

Let me know what you think. (a) Actually might make things easier to reason about (even if it's a bit more boilerplate), but it will be a pretty annoying refactor.